File: testdata/jennies/builders/map_of_builders/PythonBuilder/builders/map_of_builders.py
import typing
from ..cog import builder as cogbuilder
from ..models import map_of_builders


class Panel(cogbuilder.Builder[map_of_builders.Panel]):
    _internal: map_of_builders.Panel

    def __init__(self):
        self._internal = map_of_builders.Panel()

    def build(self) -> map_of_builders.Panel:
        """
        Builds the object.
        """
        return self._internal    
    
    def title(self, title: str) -> typing.Self:    
        self._internal.title = title
    
        return self
    


class Dashboard(cogbuilder.Builder[map_of_builders.Dashboard]):
    _internal: map_of_builders.Dashboard

    def __init__(self):
        self._internal = map_of_builders.Dashboard()

    def build(self) -> map_of_builders.Dashboard:
        """
        Builds the object.
        """
        return self._internal    
    
    def panels(self, panels: dict[str, cogbuilder.Builder[map_of_builders.Panel]]) -> typing.Self:    
        panels_resources = { key1: val1.build() for (key1, val1) in panels.items() }
        self._internal.panels = panels_resources
    
        return self
    

File: testdata/jennies/builders/map_of_builders/RustBuilder/src/builders/map_of_builders.rs
use crate::cog;
use crate::types::map_of_builders::Dashboard;
use crate::types::map_of_builders::Panel;
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct PanelBuilder {
    internal: Panel,
    errors: Vec<cog::BuildError>,
}

impl PanelBuilder {
    pub fn new() -> Self {
        Self {
            internal: Panel::default(),
            errors: Vec::new(),
        }
    }
}

impl Default for PanelBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl PanelBuilder {
    pub fn title(mut self, title: String) -> Self {
        self.internal.title = title;

        self
    }
}

impl cog::Builder<Panel> for PanelBuilder {
    fn build(&self) -> Result<Panel, Vec<cog::BuildError>> {
        if !self.errors.is_empty() {
            return Err(self.errors.clone());
        }

        Ok(self.internal.clone())
    }
}

#[derive(Debug, Clone)]
pub struct DashboardBuilder {
    internal: Dashboard,
    errors: Vec<cog::BuildError>,
}

impl DashboardBuilder {
    pub fn new() -> Self {
        Self {
            internal: Dashboard::default(),
            errors: Vec::new(),
        }
    }
}

impl Default for DashboardBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl DashboardBuilder {
    pub fn panels(mut self, panels: HashMap<String, impl cog::Builder<Panel>>) -> Self {
        let mut built0 = std::collections::HashMap::new();
        for (key0, item0) in panels {
            let built1 = match item0.build() {
                Ok(val) => val,
                Err(mut err) => {
                    self.errors.append(&mut err);
                    return self;
                }
            };
            built0.insert(key0, built1);
        }
        self.internal.panels = built0;

        self
    }
}

impl cog::Builder<Dashboard> for DashboardBuilder {
    fn build(&self) -> Result<Dashboard, Vec<cog::BuildError>> {
        if !self.errors.is_empty() {
            return Err(self.errors.clone());
        }

        Ok(self.internal.clone())
    }
}
